### PR TITLE
build SurgeXTRack with CMake

### PR DESCRIPTION
**Pull request by qno**
I added proper CMake support for compiling SurgeXTRack.

I updated the README with usage instructions. 

This solves the problem of  [Has Anyone made Code Nav tools in CLion work](https://community.vcvrack.com/t/has-anyone-made-code-nav-tools-in-clion-work/18130?u=qno) topic in a clean way.

### How does it work?

I added a `RackSDK.cmake` which basically translates all Makefile build and compiler flags into CMake. This can be used by other projects, even could become part of the Rack SDK.

Now it is possible to build all dependend libraries with the same build settings defined by Rack SDK.
It builds all surge libraries and the SurgeXTRack library with CMake.
The official plugin will still be build with Make, which uses now the SurgeXTRack library by `make dist` as usual.
I added also a target for building the plugin shared library with CMake for development purpose.

And because the complete surge dependencies are build with the same error flag settings I needed to add a lot of `-Wno-xyz` flags to surge build targets to be able to compile the dependencies now. I didn't just wanted to shut them off completely.

Let me know if this is OK or if you need more information or changes.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,42 +1,123 @@
-cmake_minimum_required(VERSION 3.20)
-set(CMAKE_OSX_ARCHITECTURES "x86_64")
-project(rack-hack VERSION 1.0 LANGUAGES C CXX)
+cmake_minimum_required(VERSION 3.16)
```
Comment (qno) [inline CMakeLists.txt:1]: Note: The rack-plugin-toolchain docker image currently only has CMake 3.16 installed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,42 +1,123 @@
...
+
+find_package(Git REQUIRED)
+
+execute_process(COMMAND git rev-parse --short HEAD WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+        OUTPUT_VARIABLE SURGE_RACK_PLUG_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+execute_process(COMMAND git rev-parse --short HEAD WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/surge
+        OUTPUT_VARIABLE SURGE_RACK_SURGE_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+set(SURGE_RACK_BASE_VERSION XT1-0-1)
```
Comment (qno) [inline CMakeLists.txt:28]: Is this version still correct?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,42 +1,123 @@
...
+
+message(STATUS "SURGE_RACK_BASE_VERSION ${SURGE_RACK_BASE_VERSION}")
+message(STATUS "SURGE_RACK_PLUG_VERSION ${SURGE_RACK_PLUG_VERSION}")
+message(STATUS "SURGE_RACK_SURGE_VERSION ${SURGE_RACK_SURGE_VERSION}")
+
```
Comment (qno) [inline CMakeLists.txt:33]: Needs something to be changed or added to the `SURGE_` settings here?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,42 +1,123 @@
...
+  set(NO_WARN_STRINGOP_TRUNCATION -Wno-string-concatenation)
+  set(NO_WARN_CLASS_MEMACCESS -Wno-class-varargs)
+  set(NO_WARN_INFINITE_RECURSION -Wno-infinite-recursion)
+  set(NO_WARN_UNUSED_PRIVATE_FIELD -Wno-unused-private-field)
+  set(NO_WARN_UNUSED_LAMBDA_CAPTURE -Wno-unused-lambda-capture)
+  set(NO_WARN_MISMATCHED_TAGS -Wno-mismatched-tags)
+  set(NO_WARN_SELF_ASSIGN -Wno-self-assign)
+endif ()
+
+target_compile_options(juce_dsp_rack_sub PRIVATE -Wno-deprecated-copy ${NO_WARN_MAYBE_UNINITIALIZED})
```
Comment (qno) [inline CMakeLists.txt:71]: Let me know if this is OK or do you would rather disable warnings completely for some dependency libs (for 3rd party libs like sqlite.)

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,42 +1,123 @@
...
-        USES_TERMINAL)
+file(GLOB LICENSE LICENSE*)
+file(INSTALL plugin.json res docs patches presets ${LICENSE} README.md DESTINATION ${PLUGIN_NAME})
+file(INSTALL surge/resources/surge-shared/configuration.xml
+             surge/resources/surge-shared/windows.wt
+             surge/resources/data/wavetables
+             surge/resources/data/wavetables_3rdparty
+             surge/resources/data/fx_presets
+             res/surge_extra_data/fx_presets
+    DESTINATION ${PLUGIN_NAME}/build/surge-data)
```
Comment (qno) [inline CMakeLists.txt:120]: Is the folder `build/surge-data` correct or shouldn't this just be `surge-data`? Is this folder used/referenced by source code?

File: RackSDK.cmake [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+if (${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+  if (NOT MINGW)
+    message(FATAL_ERROR "Rack plugin development environment is only supported for MSYS2/MinGW")
+  endif ()
+  add_compile_definitions(ARCH_WIN ARCH_X64)
+  add_compile_definitions(_USE_MATH_DEFINES)
+  # disabled - causes surge-rack/surge/libs/sst/sst-plugininfra/src/misc_windows.cpp:20:20: error: cannot convert 'const char*' to 'LPCWSTR' {aka 'const wchar_t*'}
+  # see also https://gcc.gnu.org/onlinedocs/gcc/x86-Windows-Options.html
+  #add_compile_options(-municode)
```
Comment (qno) [inline RackSDK.cmake:32]: Here something seems to be not correct set in  https://github.com/surge-synthesizer/sst-plugininfra/blob/main/src/misc_windows.cpp#L19.

File: src/SurgeXT.cpp [hunk 7]
```diff
@@ -14,7 +14,6 @@
  */
 
 #include "SurgeXT.h"
-#include "SurgeStorage.h"
```
Comment (qno) [inline src/SurgeXT.cpp:17]: This is unused include, so I removed it.

**Comment by baconpaul** [comment]
First: Thank you very much for tackling this.

I want to review this carefully and won't be able to until Friday morning most likely. The workflow i have now is really really good for me even though the cmakefile is a bit of a hack. I want to see if it still works and make adjustments carefully.

Ideally it would still work and we can merge this, or merge the modifications that result. I would be happier with the current mode not being hacky. But most important thing is I have my clion working *and* I can run from clion, update in place, and launch rack, which the hack one does so just need a bit of time.

Thank you again for all your contributions to the surge project. I really appreciate them. If you ever hop up on our discord make sure to ask me or evil to tag you as a surge team member (which makes your name go orange).

**Comment by qno** [comment]
I'm happy to help. Btw - with CLion it works like a charm now, and plugin binary is also smaller now.

**Comment by baconpaul** [comment]
OK you know what, this is too exciting. I am testing it now :)

**Comment by qno** [comment]
> awesome
> 
> you may have noticed i had an old target which just copied the plugin.dylib to my rack dir directly - I'll probably need to reinstate that
> 
> At a cursory look, it looks super smart. I just need a bit of time to test it and review it.

This should be easy, just set the `-DCMAKE_INSTALL_PREFIX` to the Rack2 plugin directory, e.g.
`cmake -DCMAKE_INSTALL_PREFIX=$HOME/Documents/Rack2/plugin -DRACK_SDK_DIR=location-of-rack-SDK-or-source -DCMAKE_OSX_ARCHITECTURES="x86_64" ../surge-rack`

Then hit Install in CLion menu:
image

**Comment by baconpaul** [comment]
ahh yeah but won't that install the assets also? That's super useful but a 'just copy the plugin.dylib and run rack' is what I really need.

I also note the newly added `SURGE_SKIP_ODDSOUND_MTS` got dropped. Would you rather I merge and fix later, or would you like to add that?

**Comment by qno** [comment]
I can do this. Also the assets are not installed again when CMake detects no changes on them.

**Comment by baconpaul** [comment]
On my mac I also get these warnings

```
[5/94] Building CXX object CMakeFiles/plugin.dir/src/SurgeXT.cpp.o
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fuse-ld=gold' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-MD' [-Wunused-command-line-argument]
[69/94] Building CXX object CMakeFiles/SurgeXTRack.dir/src/DelayLineByFreq.cpp.o
clang: warning: argument unused during compilation: '-static-libstdc++' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-MD' [-Wunused-command-line-argument]
```

expected?

**Comment by baconpaul** [comment]
but dinner time here. More later and if not more tomorrow.
Very very promising

**Comment by qno** [comment]
I can clean the flags up, I don't care normally if it is not an error.

Here it is 0:48. So I would need to go to bed now.

**Comment by baconpaul** [comment]
OK this is definitely almost exactly what we want so I'm going to merge it and we can add and tweak as incremental.

First thing I'll do is add a little 'plugin-installed-alone' target which will let me have one key to clion

seeing all the warnings in clion is great, as is having an arm and debug build option. This is fantastic work. Thank you very much.

**Comment by baconpaul** [comment]
Ahh that was premature. the dylib this makes doesn't actually work on my mac and contains none of the actual code! 

Lemme see if I can debug if not I'll revert it and we can look again.

**Comment by baconpaul** [comment]
OK great got it working. Will do another push shortly.

**Pull request merged**